sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryHandler.java

```diff
@@ -537,8 +536,7 @@ protected PlannerResult planWithDruidConvention() throws ValidationException
   @Nullable
   private RelRoot possiblyWrapRootWithOuterLimitFromContext(RelRoot root)
   {
-    Object outerLimitObj = handlerContext.queryContext().get(PlannerContext.CTX_SQL_OUTER_LIMIT);
-    Long outerLimit = DimensionHandlerUtils.convertObjectToLong(outerLimitObj, true);
+    Long outerLimit = handlerContext.queryContext().getLong(PlannerContext.CTX_SQL_OUTER_LIMIT);
```

Comment:
You've identified another nasty inconsistency.

Our general rule for context values seems to be that they should be in JSON-compatible format, since we use Jackson to deserialize. So, if this context value appears in the context as a JSON number, it must follow the JSON rules and will be converted to `Integer`, `Long`, `Float`, etc. depending on the format of the number. The usual `QueryContexts.getAsLong()` method seems to convert from whatever Jackson created to get an actual `Long`. This handles `x` where `x` is a number.

We also handle `"x"` where we quote the number from above. JSON says that a string is a string: there is no JSON-defined conversion from strings to numbers. If we want to allow that, then Druid has to provide the logic to do so. Unfortunately, before your fix (and my doubling-down on your approach), it looked like every bit of code could make up its own parsing rules.

We can perhaps standardize and say that, if we want to parse `{"foo": "x"}` as a number, then we want the same result as if the user had actually written `{"foo": x}`, where 'x` is a valid JSON number. This means that, when the value is a string, what we actually want to do to parse is:

```java
Long longValue = objectMapper.readValue(context.getString("x"), Long.class);
```

We probably don't actually want to do the above: it is costly and requires a mapper. But, we want our string-to-long conversion to produce the same result as the above. A new commit added tests to probe the odd cases to see how closely the context functions follow JSON and the dimension utilities.

As it turns out, both the dimension utilities and Jackson does handle a conversion from "20.0" to a Long, but the context `getAsLong()` does not. To fix this, I added the `convertObjectToLong()` code to `getAsLong()` so that the context methods also handle this case. That fix should address the concern you raised.